src/meta_schedule/feature_extractor/per_block_feature.cc

```diff
@@ -254,16 +254,192 @@ Feature::ForKindFeature::ForKindFeature(const ForVec& loops) {
+      for (int i = 0; i < n_samples; ++i) {
+        arith_intensity_curve[i] = 0.0;
+      }
+      return;
+    }
+    total_compute_ops = compute_ops.back();  // i.e. total_compute_ops = log2(total_compute_ops)
+    int p = 0;
+    for (int i = 0; i < n_samples; ++i) {
+      double& result = arith_intensity_curve[i];
+      double cur_compute_ops = static_cast<double>(i + 1) / n_samples * total_compute_ops;
+      // Find the first `p` that `compute[p] >= total * (i + 1) / N`
+      for (; p < n_loops; ++p) {
+        if (compute_ops[p] >= cur_compute_ops - 1e-4) {
+          break;
+        }
+      }
+      CHECK_LT(p, n_loops);
+      if (p == 0) {
+        result = compute_ops[p] / memory_bytes[p];
+      } else {
+        double base = compute_ops[p - 1] / memory_bytes[p - 1];
+        double slope =
+            (compute_ops[p] / memory_bytes[p] - compute_ops[p - 1] / memory_bytes[p - 1]) /
+            (compute_ops[p] - compute_ops[p - 1]);
+        result = base + slope * (cur_compute_ops - compute_ops[p - 1]);
+      }
+    }
+  }
+};
+
+}  // namespace group3
+
+namespace group4 {
+
+/*! \brief Group 4 feature */
+struct Feature {
+  // Since allocated buffers can be in local, shared, or global memory, each attribute below is
+  // divided into three cases.
+  // The size of allocated buffer in bytes
+  int64_t alloc_size_local = 0;
+  int64_t alloc_size_shared = 0;
+  int64_t alloc_size_global = 0;
+  // alloc_outer_prod * alloc_inner_prod
+  int64_t alloc_prod_local = 0;
+  int64_t alloc_prod_shared = 0;
+  int64_t alloc_prod_global = 0;
+  // The product of lengths of loops outside the scope of the alloc
+  int64_t alloc_outer_prod_local = 1;
+  int64_t alloc_outer_prod_shared = 1;
+  int64_t alloc_outer_prod_global = 1;
+
+  static constexpr int64_t kCount = 12;
+
+  void Export(std::vector<double>* v, int64_t outer_prod) const {
+    double vs[] = {
+        slog(alloc_size_local),
+        slog(alloc_size_shared),
+        slog(alloc_size_global),
+        slog(alloc_prod_local),
+        slog(alloc_prod_shared),
+        slog(alloc_prod_global),
+        slog(alloc_outer_prod_local),
+        slog(alloc_outer_prod_shared),
+        slog(alloc_outer_prod_global),
+        slog(static_cast<double>(outer_prod) / alloc_outer_prod_local),
+        slog(static_cast<double>(outer_prod) / alloc_outer_prod_shared),
+        slog(static_cast<double>(outer_prod) / alloc_outer_prod_global),
+    };
+    v->insert(v->end(), std::begin(vs), std::end(vs));
+  }
+
+  Feature() = default;
+
+  explicit Feature(const LoopNest& loop_nest, const BlockRealizeNode* realize,
+                   arith::Analyzer* analyzer) {
+    for (const Buffer& buffer : realize->block->alloc_buffers) {
+      std::vector<int64_t> shape = utils::GetBufferShape(buffer, analyzer);
+      int64_t numel = 1;
+      for (int64_t x : shape) {
+        numel *= x;
+      }
+      const runtime::StorageScope storage_scope = runtime::StorageScope::Create(buffer.scope());
```

Comment:
nit: no need for the const qualifier. the compiler could identify constants and do constant-related optimization automatically.

```suggestion
     runtime::StorageScope storage_scope = runtime::StorageScope::Create(buffer.scope());
```

Reply:
Addressed :)